File: mySolutions/42. Trapping Rain Water/42. Trapping Rain Water.3.py
from collections import deque

class Solution:
    def trap(self, height: List[int]) -> int:
        
        sum = 0
        stack = deque()
        i = 0
        while i <= len(height) - 1:
          rightHi = height[i]
          
          if len(stack) <= 1 or height[stack[-1]] >= rightHi:
            # print ("rightHi",rightHi, "stack[-1]", stack[-1] if stack else 0, stack)
            stack.append(i)
            i += 1
          else:
            midHi = height[stack.pop()]
            left = stack[-1] 
            leftHi = height[left]
            vol = (i - left - 1) * (min(rightHi, leftHi) - midHi)
            # print ("right", i, "left", left, "rightHi", rightHi, "midHi", midHi, "leftHi", leftHi, "vol", vol)
            
            sum += vol if vol >= 0 else 0
      
        return sum 

File: mySolutions/42. Trapping Rain Water/42. Trapping Rain Water.3.java
class Solution {
    public int trap(int[] height) {
        
        Stack<Integer> st = new Stack<>();
        
        int index = 0;
        int total = 0;
        while (index <= height.length - 1) {
            int rH = height[index];
            
            if (st.size() == 0 || height[st.peek()] > rH) {
                st.push(index);
                index += 1;
            } else {
                int bot = st.pop();
                if (st.size() == 0) {
                    st.push(index);
                    index += 1;
                    continue;
                }
                int left = st.peek();
                total += (index - left - 1) * (Math.min(height[index], height[left]) - height[bot]);
            }
        }
        return total;        
    }
}